File: main/check-completeness-of-a-binary-tree/check-completeness-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static bool isCompleteTree(const TreeNode* root) noexcept;
};

bool Solution::isCompleteTree(const TreeNode* root) noexcept
{
    queue<const TreeNode*> q;
    
    for (q.push(root); q.front(); q.pop()) {
        q.push(q.front()->left);
        q.push(q.front()->right);
    }
    
    for (; !q.empty(); q.pop())
        if (q.front()) return false;
    
    return true;
}

File: main/check-completeness-of-a-binary-tree/check-completeness-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isCompleteTree(TreeNode root) {
        final TreeNode sentinel = new TreeNode(0);
        
        final Queue<TreeNode> nodes = new ArrayDeque<>();
        nodes.add(root == null ? sentinel : root);
        
        while ((root = nodes.remove()) != sentinel) {
            nodes.add(root.left == null ? sentinel : root.left);
            nodes.add(root.right == null ? sentinel : root.right);
        }
        
        while (!nodes.isEmpty())
            if (nodes.remove() != sentinel) return false;
        
        return true;
    }
}
